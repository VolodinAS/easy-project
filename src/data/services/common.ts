import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { getConfigValue } from '@ijl/cli';
import { Author, CommonResponse, NewProject, NewTask, ProjectId, ProjectType, TaskGetByIdType, TaskId, TaskType } from '../models/common';
import { authToken } from '../../utils/token';

export const apiCommon = createApi({
    reducerPath: 'commonApi',
    baseQuery: fetchBaseQuery({
        baseUrl: getConfigValue('easy-project.api-base-url'),
        prepareHeaders: headers => {
            if (authToken.token) {
                headers.set('Authorization', `Bearer ${authToken.token}`);
            }
            return headers;
        },
    }),
    tagTypes: ['ProjectType', 'TaskType'],
    endpoints: (build) => ({

        getProjects: build.query<CommonResponse<ProjectType[]>, void>({
            query: () => ({
                url: '/projects'
            }),
            providesTags: result => ['ProjectType']
        }),
        getProjectById: build.query<CommonResponse<ProjectType>, { projectId: string }>({
            query: ({projectId}) => ({
                url: `/projects/${projectId}`
            })
        }),
        newProject: build.mutation<CommonResponse<NewProject>, {title: string, code: string, members: object}>({
            query: ({ title, code, members }) => ({
                url: '/projects/new',
                method: 'post',
                body: { title, code, members }
            }),
            invalidatesTags: ['ProjectType']
        }),
        editProject: build.mutation<CommonResponse<NewProject>, {projectId: string, title: string, code: string, members: object}>({
            query: ({ projectId, title, code, members }) => ({
                url: '/projects/edit',
                method: 'post',
                body: { projectId, title, code, members }
            }),
            invalidatesTags: ['ProjectType']
        }),
        deleteProject: build.mutation<CommonResponse<ProjectId>, {projectId: string}>({
            query: (projectId) => ({
                url: '/projects/delete',
                method: 'post',
                body: { projectId }
            }),
            invalidatesTags: ['ProjectType']
        }),



        getTasks: build.query<CommonResponse<TaskType[]>, { projectId: string }>({
            query: ({projectId}) => ({
                url: `/tasks/${projectId}`
            }),
            providesTags: result => ['TaskType']
        }),
        getTaskById: build.query<CommonResponse<TaskType>, { projectId: string, taskId : string}>({
            query: ({ projectId, taskId }) => ({
                url: `/tasks/${projectId}/${taskId}`
            }),
            providesTags: result => ['TaskType']
        }),
        newTask: build.mutation<CommonResponse<NewTask>, { projectId: string, title: string, description: string, type: number, status: number }>({
            query: ({ projectId, title, description, type, status }) => ({
                url: '/tasks/create',
                method: 'post',
                body: { projectId, title, description, type, status }
            }),
            invalidatesTags: ['TaskType']
        }),
        editTask: build.mutation<CommonResponse<NewTask>, { projectId: string, taskId: string, title: string, description: string, type: number, status: number }>({
            query: ({ projectId, taskId, title, description, type, status }) => ({
                url: '/tasks/edit',
                method: 'post',
                body: { projectId, taskId, title, description, type, status }
            }),
            invalidatesTags: ['TaskType']
        }),
        deleteTask: build.mutation<CommonResponse<TaskId>, { projectId: string, taskId: string }>({
            query: ({ projectId, taskId }) => ({
                url: '/tasks/delete',
                method: 'post',
                body: { projectId, taskId }
            }),
            invalidatesTags: ['TaskType']
        }),


        getUsersForSelect: build.query<CommonResponse<Author[]>, void>({
            query: () => ({
                url: '/users/getForSelect',
            }),
        })
    }),
});

export const {
    useGetProjectsQuery,
    useGetProjectByIdQuery,
    useLazyGetProjectByIdQuery,
    useNewProjectMutation,
    useEditProjectMutation,
    useDeleteProjectMutation,
    useGetTasksQuery,
    useGetTaskByIdQuery,
    useLazyGetTaskByIdQuery,

    useNewTaskMutation,
    useEditTaskMutation,
    useDeleteTaskMutation,

    useGetUsersForSelectQuery,
    useLazyGetUsersForSelectQuery,
} = apiCommon;