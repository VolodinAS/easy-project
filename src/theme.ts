const theme = {
    color: {
        typeTask: '#F8BFBF',
        typeFeature: '#B7CFFF',
        typeBug: '#FEFBBD',
        black: '#000000',
        lightPurple: '#CFC7D9',
        purple: '#7A52AA',
        lightGray: '#F5F5F5',
        darkGray: '#9C9C9C',
        red: '#EB0C0C',
        white: '#FFFFFF',
        smallText: '#999999',
        opacityGray: '#ACACAC',
        landingColors: {
            mainGray: '#D5E6EA',
            mainBlue: '#3848A9',
            benefitsItemBackground: '#EDD8BE',
            buttonStartBackground: '#E43958',
            footerBackground: '#343F53',
            headerLogo: '#010101',
            headerBackground: '#F1E7DA'
        },
        dashboardColors: {
            overlayButtonCreate: '#ABE0E7',
            overlayButtonYes: '#B4E7AB',
            overlayButtonNo: '#F9CCBE',
        }
    }
};

export default theme;
