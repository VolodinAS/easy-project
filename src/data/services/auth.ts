import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { getConfigValue } from '@ijl/cli';

import { CommonResponse } from '../models/common';
import { SignInData } from '../models/sing-in';

export const api = createApi({
    reducerPath: 'authApi',
    baseQuery: fetchBaseQuery({
        baseUrl: getConfigValue('easy-project.api-base-url'),
    }),
    tagTypes: [],
    endpoints: (build) => ({
        signIn: build.mutation<CommonResponse<SignInData>, {email: string, password: string}>({
            query: ({ email, password }) => ({
                url: '/sign-in',
                method: 'post',
                body: { email, password }
            })
        }),

        signUp: build.mutation<CommonResponse<unknown>, {email: string, password: string, login: string, name_secondname: string}>({
            query: ({ email, password, login, name_secondname }) => ({
                url: '/registration',
                method: 'post',
                body: { email, password, login, name_secondname }
            })
        })
    }),


});

export const { useSignInMutation, useSignUpMutation } = api;