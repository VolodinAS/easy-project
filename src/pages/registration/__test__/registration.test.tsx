/* eslint-disable check-file/folder-naming-convention */
import React from 'react';
import { describe, it, expect, jest, beforeAll, afterAll } from '@jest/globals';
import { render, waitFor, screen, fireEvent, getAllByText, getByText } from '@testing-library/react';
import { setupServer } from 'msw/node';
import { rest } from 'msw';
import { ApiProvider } from '@reduxjs/toolkit/dist/query/react';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from '@emotion/react';

import 'whatwg-fetch';

import theme from '../../../theme';
import { api } from '../../../data/services/auth';
import Registration from '../index';

import signUpDataSuccess from '../../../../stubs/json/signup/success.json';
import signUpDataExistEmail from '../../../../stubs/json/signup/error-exist-email.json';
import signUpDataExistLogin from '../../../../stubs/json/signup/error-exist-login.json';

const serverSuccess = setupServer(
    rest.post('/api/registration', (req, res, ctx) => {
        return res(
            ctx.json(signUpDataSuccess)
        );
    })
);

const serverExistEmail = setupServer(
    rest.post('/api/registration', (req, res, ctx) => {
        return res(
            ctx.status(400),
            ctx.json(signUpDataExistEmail)
        );
    })
);

const serverExistLogin = setupServer(
    rest.post('/api/registration', (req, res, ctx) => {
        return res(
            ctx.status(400),
            ctx.json(signUpDataExistLogin)
        );
    })
);

describe('Testing Registration Component COMMON', () => {
    beforeAll( () => {
        serverSuccess.listen();
    } );
    afterAll( () => {
        serverSuccess.close();
    } );
    it('check render', () => {
        const { container } = render(
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                    <ApiProvider api={api}>
                        <Registration/>
                    </ApiProvider>
                </BrowserRouter>
            </ThemeProvider>
        );
        expect(container).toMatchSnapshot();
    });

    it('submit empty form', async () => {
        const errorEmpty = 'Поле не должно быть пустым';
        const { getByTestId, getByLabelText } = render(
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                    <ApiProvider api={api}>
                        <Registration/>
                    </ApiProvider>
                </BrowserRouter>
            </ThemeProvider>
        );
        const signUpForm = getByTestId('signup-form');

        fireEvent.submit(signUpForm);

        await waitFor(() => {
            const emptyLabel = getAllByText(signUpForm, errorEmpty)[0] as HTMLParagraphElement;
            expect(emptyLabel.textContent).toContain(errorEmpty);
        });

        expect(signUpForm).toMatchSnapshot();
    });
});

describe('Testing Registration Component EXIST EMAIL', () => {
    beforeAll( () => {
        serverExistEmail.listen();
    } );
    afterAll( () => {
        serverExistEmail.close();
    } );

    it('submit exist email', async () => {
        const errorEmpty = 'Email уже занят';
        const { getByTestId, getByLabelText } = render(
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                    <ApiProvider api={api}>
                        <Registration/>
                    </ApiProvider>
                </BrowserRouter>
            </ThemeProvider>
        );

        fireEvent.change(getByLabelText(/Почта/i), {
            target: { value: 'alex_vovkin@mail.ru' }
        });

        fireEvent.change(getByLabelText(/Логин/i), {
            target: { value: 'some login' }
        });

        fireEvent.change(getByLabelText(/Фамилия Имя/i), {
            target: { value: 'some fio' }
        });

        fireEvent.change(getByLabelText(/Пароль/i), {
            target: { value: 'some password' }
        });

        const signUpForm = getByTestId('signup-form');

        fireEvent.submit(signUpForm);

        await waitFor(() => {
            const emptyLabel = getByText(document.body, errorEmpty);
            expect(emptyLabel.innerHTML).toContain(errorEmpty);
        });

        expect(signUpForm).toMatchSnapshot();
    });
});

describe('Testing Registration Component EXIST LOGIN', () => {
    beforeAll( () => {
        serverExistLogin.listen();
    } );
    afterAll( () => {
        serverExistLogin.close();
    } );

    it('submit exist login', async () => {
        const errorEmpty = 'Логин уже занят';
        const { getByTestId, getByLabelText } = render(
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                    <ApiProvider api={api}>
                        <Registration/>
                    </ApiProvider>
                </BrowserRouter>
            </ThemeProvider>
        );

        fireEvent.change(getByLabelText(/Почта/i), {
            target: { value: 'testing@test.test' }
        });

        fireEvent.change(getByLabelText(/Логин/i), {
            target: { value: 'VolodinAS' }
        });

        fireEvent.change(getByLabelText(/Фамилия Имя/i), {
            target: { value: 'some fio' }
        });

        fireEvent.change(getByLabelText(/Пароль/i), {
            target: { value: 'some password' }
        });

        const signUpForm = getByTestId('signup-form');

        fireEvent.submit(signUpForm);

        await waitFor(() => {
            const emptyLabel = getByText(document.body, errorEmpty);
            expect(emptyLabel.innerHTML).toContain(errorEmpty);
        });

        expect(signUpForm).toMatchSnapshot();
    });
});