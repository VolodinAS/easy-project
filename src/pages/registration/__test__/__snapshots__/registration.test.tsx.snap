// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Registration Component COMMON check render 1`] = `
<div>
  <div
    class="css-1wcmp47"
  >
    <section
      class="css-1prn45"
    >
      <button
        class="css-1640pth"
        title="language select"
        value="ru"
      >
        <img
          class="css-zlc08d"
          src="test-file-stub"
        />
      </button>
    </section>
    <div
      class="css-1sw1t6e"
    >
      <p
        class="css-a8wdq9"
      >
        Создай свой аккаунт
      </p>
      <form
        data-testid="signup-form"
        novalidate=""
      >
        <fieldset
          class="css-1nvs5py"
        >
          <div
            class="css-ahns2k"
          >
            <label
              class="css-12e7qyv"
            >
              <span
                class="css-fxi74e"
              >
                Логин
                *
              </span>
              <input
                class="css-d17r1a"
                name="login"
                placeholder="Введите логин..."
                required=""
                type="text"
                value=""
              />
            </label>
          </div>
          <div
            class="css-ahns2k"
          >
            <label
              class="css-12e7qyv"
            >
              <span
                class="css-fxi74e"
              >
                Фамилия Имя
                *
              </span>
              <input
                class="css-d17r1a"
                name="name_secondname"
                placeholder="Фамилия Имя..."
                required=""
                type="Name_SecondName"
                value=""
              />
            </label>
          </div>
          <div
            class="css-ahns2k"
          >
            <label
              class="css-12e7qyv"
            >
              <span
                class="css-fxi74e"
              >
                Почта
                *
              </span>
              <input
                class="css-d17r1a"
                name="email"
                placeholder="Введите почту..."
                required=""
                type="email"
                value=""
              />
            </label>
          </div>
          <div
            class="css-ahns2k"
          >
            <label
              class="css-12e7qyv"
            >
              <span
                class="css-fxi74e"
              >
                Пароль
                *
              </span>
              <input
                autocomplete="off"
                class="css-d17r1a"
                name="password"
                placeholder="Введите пароль..."
                required=""
                type="password"
                value=""
              />
            </label>
          </div>
        </fieldset>
        <div
          class="css-jtawwa"
        >
          <p>
            Уже есть аккаунт?
            <a
              class="css-xyki2o"
              href="/easy-project/sign-in"
              title="Войти если есть аккаунт"
            >
               Войти
            </a>
          </p>
        </div>
        <div
          class="css-889chh"
        >
          <button
            class="css-ddydcb"
            type="submit"
          >
            Присоединиться
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`Testing Registration Component COMMON submit empty form 1`] = `
<form
  data-testid="signup-form"
  novalidate=""
>
  <fieldset
    class="css-1nvs5py"
  >
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Логин
          *
        </span>
        <input
          class="css-d17r1a"
          name="login"
          placeholder="Введите логин..."
          required=""
          type="text"
          value=""
        />
      </label>
      <p
        class="css-11un70f"
      >
        Поле не должно быть пустым
      </p>
    </div>
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Фамилия Имя
          *
        </span>
        <input
          class="css-d17r1a"
          name="name_secondname"
          placeholder="Фамилия Имя..."
          required=""
          type="Name_SecondName"
          value=""
        />
      </label>
      <p
        class="css-11un70f"
      >
        Поле не должно быть пустым
      </p>
    </div>
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Почта
          *
        </span>
        <input
          class="css-d17r1a"
          name="email"
          placeholder="Введите почту..."
          required=""
          type="email"
          value=""
        />
      </label>
      <p
        class="css-11un70f"
      >
        Поле не должно быть пустым
      </p>
    </div>
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Пароль
          *
        </span>
        <input
          autocomplete="off"
          class="css-d17r1a"
          name="password"
          placeholder="Введите пароль..."
          required=""
          type="password"
          value=""
        />
      </label>
      <p
        class="css-11un70f"
      >
        Поле не должно быть пустым
      </p>
    </div>
  </fieldset>
  <div
    class="css-jtawwa"
  >
    <p>
      Уже есть аккаунт?
      <a
        class="css-xyki2o"
        href="/easy-project/sign-in"
        title="Войти если есть аккаунт"
      >
         Войти
      </a>
    </p>
  </div>
  <div
    class="css-889chh"
  >
    <button
      class="css-ddydcb"
      type="submit"
    >
      Присоединиться
    </button>
  </div>
</form>
`;

exports[`Testing Registration Component EXIST EMAIL submit exist email 1`] = `
<form
  data-testid="signup-form"
  novalidate=""
>
  <fieldset
    class="css-1nvs5py"
  >
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Логин
          *
        </span>
        <input
          class="css-d17r1a"
          name="login"
          placeholder="Введите логин..."
          required=""
          type="text"
          value="some login"
        />
      </label>
      
    </div>
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Фамилия Имя
          *
        </span>
        <input
          class="css-d17r1a"
          name="name_secondname"
          placeholder="Фамилия Имя..."
          required=""
          type="Name_SecondName"
          value="some fio"
        />
      </label>
      
    </div>
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Почта
          *
        </span>
        <input
          class="css-d17r1a"
          name="email"
          placeholder="Введите почту..."
          required=""
          type="email"
          value="alex_vovkin@mail.ru"
        />
      </label>
      
    </div>
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Пароль
          *
        </span>
        <input
          autocomplete="off"
          class="css-d17r1a"
          name="password"
          placeholder="Введите пароль..."
          required=""
          type="password"
          value="some password"
        />
      </label>
      
    </div>
  </fieldset>
  <div
    class="css-jtawwa"
  >
    <p>
      Уже есть аккаунт?
      <a
        class="css-xyki2o"
        href="/easy-project/sign-in"
        title="Войти если есть аккаунт"
      >
         Войти
      </a>
    </p>
  </div>
  <div
    class="css-889chh"
  >
    <button
      class="css-ddydcb"
      type="submit"
    >
      Присоединиться
    </button>
  </div>
</form>
`;

exports[`Testing Registration Component EXIST LOGIN submit exist login 1`] = `
<form
  data-testid="signup-form"
  novalidate=""
>
  <fieldset
    class="css-1nvs5py"
  >
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Логин
          *
        </span>
        <input
          class="css-d17r1a"
          name="login"
          placeholder="Введите логин..."
          required=""
          type="text"
          value="VolodinAS"
        />
      </label>
      
    </div>
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Фамилия Имя
          *
        </span>
        <input
          class="css-d17r1a"
          name="name_secondname"
          placeholder="Фамилия Имя..."
          required=""
          type="Name_SecondName"
          value="some fio"
        />
      </label>
      
    </div>
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Почта
          *
        </span>
        <input
          class="css-d17r1a"
          name="email"
          placeholder="Введите почту..."
          required=""
          type="email"
          value="testing@test.test"
        />
      </label>
      
    </div>
    <div
      class="css-ahns2k"
    >
      <label
        class="css-12e7qyv"
      >
        <span
          class="css-fxi74e"
        >
          Пароль
          *
        </span>
        <input
          autocomplete="off"
          class="css-d17r1a"
          name="password"
          placeholder="Введите пароль..."
          required=""
          type="password"
          value="some password"
        />
      </label>
      
    </div>
  </fieldset>
  <div
    class="css-jtawwa"
  >
    <p>
      Уже есть аккаунт?
      <a
        class="css-xyki2o"
        href="/easy-project/sign-in"
        title="Войти если есть аккаунт"
      >
         Войти
      </a>
    </p>
  </div>
  <div
    class="css-889chh"
  >
    <button
      class="css-ddydcb"
      type="submit"
    >
      Присоединиться
    </button>
  </div>
</form>
`;
